### ISSUE TITLE
hoisted $inject assignment can break inheritance-using TypeScript-code

### ISSUE THREAD

**Issue by edgar0011** [olov/ng-annotate#214]
when JS generated from modern ES6 like langs, Babel, Typescript, using OOP classes, super calls, $inject is suddenly used too soon, befor atuall extends older version, 1.0.2 works ok:

```js
var UILangs = (function (_super) {
        UILangs.$inject = ["$translateProvider", "$translatePartialLoaderProvider", "SERVICES_URL_PREFIX"];
        __extends(UILangs, _super);
        /**
         * Class instance constructor.
         *
         * @param $translateProvider {ITranslateProvider}
         * @param $translatePartialLoaderProvider {ITranslatePartialLoaderService}
         * @param SERVICES_URL_PREFIX {string}
         *
         * @ngInject
         */
        function UILangs($translateProvider, $translatePartialLoaderProvider, SERVICES_URL_PREFIX) {
            _super.call(this, $translateProvider, SERVICES_URL_PREFIX);
            this.$translateProvider = $translateProvider;
            this.$translatePartialLoaderProvider = $translatePartialLoaderProvider;
            this.SERVICES_URL_PREFIX = SERVICES_URL_PREFIX;
            // adding module specific localization part
            $translatePartialLoaderProvider.addPart('UI');
        }
        return UILangs;
})(general.Abstracts.Langs);
exports.UILangs = UILangs;
```

**Comment by olov** [comment]
The hoisted `$inject`-assignment was an intended change with ng-annotate 1.1 (see https://github.com/olov/ng-annotate/issues/168). Is the problem here that `__extends(UILangs, _super)` will copy `UILangs.$inject = _super.$inject` which used to not be a problem because shortly after the property was overwritten with `UILangs.$inject = ["$translateProvider" ..]` but now it happens in the wrong order?

**Comment by olov** [comment]
As a workaround until we've figured out what to do about this, just include the `UILangs.$inject = ["$translateProvider" ..` line in your source code and ng-annotate will detect that it exists and not fiddle with it.

**Comment by olov** [comment]
Another workaround you can do is to define your own `__extends` function that works like the default TypeScripe one except that it doesn't transfer the `$inject` property.

A new version of ng-annotate will most likely detect the presence of a `__extends` function-call and move the `$inject`-array below it, whenever possible.

**Comment by olov** [comment]
released fix in 1.2.0

**Comment by edgar0011** [comment]
thanks a lot, great work

**Comment by andreigabreanu** [comment]
This seems to break current code base for me:

 TypeError: Cannot set property '$inject' of undefined
```javascript
var Foo = (function (_super) {
	    __extends(Foo, _super);
	    /**
	     * @ngInject
	     * @param props
	     * @param translate
	     */
	    Foo.$inject = ["props", "translate"];
	    Foo.onEnter.$inject = ["callback"];
            ....
 
	    /**
	     * @ngInject
	     * @param callback
	     */
	    Foo.onEnter = function (callback) {
	        callback();
	    };
});
```

**Comment by iliarogov** [comment]
same problem as @andreigabreanu.

**Comment by Realgomu** [comment]
same problem used in typescript

**Comment by olov** [comment]
Fixed in 1.2.1.
